Improve command codestyle

JRoy opened the pull request:
Clean up all the commands, adds a bunch of helper methods, fixes a few bugs.

Didn't clean commands which are touched in my other pull requests

Fixes #3579
Fixes #3336
Closes #3339
Closes #3087

JRoy committed:
Everything up to eco exclusive

JRoy committed:
Merge branch '2.x' into feature/command-rewrite

JRoy committed:
Allow specifying a consumer in loop commands

JRoy committed:
Remove need for updatePlayer method

JRoy committed:
Allow specifying a consumer in loop commands

JRoy committed:
Remove args from loopOfflinePlayers

JRoy committed:
Allow all exceptions to be thrown in the loop consumer

JRoy committed:
Cleanup eco command

JRoy committed:
Merge branch '2.x' into feature/command-rewrite

JRoy committed:
Merge branch '2.x' into feature/command-rewrite

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandback.java:
```diff
@@ -32,28 +29,17 @@ protected void run(Server server, User user, String commandLabel, String[] args)
-            sender.sendMessage(tl("backOther", player.getName()));
-            teleportBack(sender, ess.getUser(player));
-        }
+    private void parseCommand(Server server, CommandSource sender, String[] args) throws Exception {
```
```suggestion
    private void parseOthers(Server server, CommandSource sender, String[] args) throws Exception {
```

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/EssentialsLoopCommand.java:
```diff
@@ -19,27 +19,31 @@ public EssentialsLoopCommand(String command) {
         super(command);
     }
 
-    protected void loopOfflinePlayers(final Server server, final CommandSource sender, final boolean multipleStringMatches, boolean matchWildcards, final String searchTerm, final String[] commandArgs) throws PlayerNotFoundException, NotEnoughArgumentsException, PlayerExemptException, ChargeException, MaxMoneyException {
```
Not sure about this; sure, this is a lot of exceptions in the method signature, but this encourages throwing whatever you want and not caring about what you catch.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/EssentialsLoopCommand.java:
```diff
@@ -19,27 +19,31 @@ public EssentialsLoopCommand(String command) {
+        loopOfflinePlayersConsumer(server, sender, multipleStringMatches, matchWildcards, searchTerm, user -> updatePlayer(server, sender, user, commandArgs));
+    }
+
+    protected void loopOfflinePlayersConsumer(final Server server, final CommandSource sender, final boolean multipleStringMatches, boolean matchWildcards, final String searchTerm, UserConsumer userConsumer) throws Exception {
```
This is an overloaded method, why change the name?

JRoy replied:
Some uses of loop methods provide null to the args array which would create an ambiguous method call due to the fact that the `String[] args` argument isn't in `loopOfflinePlayersConsumer`

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/EssentialsLoopCommand.java:
```diff
@@ -48,19 +52,23 @@ protected void loopOfflinePlayers(final Server server, final CommandSource sende
             final User user = getPlayer(server, searchTerm, true, true);
-            updatePlayer(server, sender, user, commandArgs);
+            userConsumer.accept(user);
         }
     }
 
-    protected void loopOnlinePlayers(final Server server, final CommandSource sender, final boolean multipleStringMatches, boolean matchWildcards, final String searchTerm, final String[] commandArgs) throws PlayerNotFoundException, NotEnoughArgumentsException, PlayerExemptException, ChargeException, MaxMoneyException {
+    protected void loopOnlinePlayers(final Server server, final CommandSource sender, final boolean multipleStringMatches, boolean matchWildcards, final String searchTerm, final String[] commandArgs) throws Exception {
+        loopOnlinePlayersConsumer(server, sender, multipleStringMatches, matchWildcards, searchTerm, user -> updatePlayer(server, sender, user, commandArgs));
+    }
+
+    protected void loopOnlinePlayersConsumer(final Server server, final CommandSource sender, final boolean multipleStringMatches, boolean matchWildcards, final String searchTerm, UserConsumer userConsumer) throws Exception {
```
Same concerns as above.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/EssentialsLoopCommand.java:
```diff
@@ -100,19 +108,21 @@ protected void loopOnlinePlayers(final Server server, final CommandSource sender
     }
 
-    protected abstract void updatePlayer(Server server, CommandSource sender, User user, String[] args) throws NotEnoughArgumentsException, PlayerExemptException, ChargeException, MaxMoneyException;
+    protected void updatePlayer(Server server, CommandSource sender, User user, String[] args) throws NotEnoughArgumentsException, PlayerExemptException, ChargeException, MaxMoneyException {
```
I'd prefer keeping this abstract, it's still used in most places, and the few cases where it's unused we can implement the empty method.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/EssentialsLoopCommand.java:
```diff
@@ -129,4 +139,8 @@ protected void loopOnlinePlayers(final Server server, final CommandSource sender
         return players;
     }
+
+    public interface UserConsumer {
```
Is this interface really necessary? Why not `Consumer<User>`?

JRoy replied:
Consumers are not allowed to throw exceptions

JRoy committed:
Reduce exceptions able to be thrown + make updatePlayer abstract

JRoy committed:
Update Essentials/src/com/earth2me/essentials/commands/Commandback.java

Co-authored-by: MD <1917406+md678685@users.noreply.github.com>

JRoy committed:
Fix build

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandback.java:
```diff
@@ -23,7 +20,7 @@ public Commandback() {
         CommandSource sender = user.getSource();
         if (args.length > 0 && user.isAuthorized("essentials.back.others")) {
-            this.parseCommand(server, sender, args, true);
+            this.parseCommand(server, sender, args);
```
```suggestion
            this.parseOthers(server, sender, args);
```

Fixes compile

JRoy committed:
Cleanup /enchant and /enderchest

JRoy committed:
Add CommandSource#getUser convenience method

JRoy committed:
Move User#getDisplayName to IUser interface

JRoy committed:
Cleanup xp command

JRoy committed:
Use enum in tab completion

JRoy committed:
Commandext

JRoy committed:
Exp, fireball, firework, fly

JRoy committed:
Make gamemode command a loop command

JRoy committed:
Add CommandSource#isAuthorized helper method

JRoy committed:
Convert give to loop command

JRoy committed:
Clean up god to kill commands inclusive

JRoy committed:
Clean up kit to motd commands inclusive

JRoy committed:
Clean up msg to pweather

time and weather need to be converted to loop commands because HOLY FUCK LOOK AT THAT CODE i have no brain power right now

JRoy committed:
Remove duplicate permission in plugin.yml

JRoy committed:
Remove player related code from 1.6.4

JRoy committed:
Add CommandSource#getSelfSelector helper method

JRoy committed:
Refactor ptime and pweather

These were the worst commands i've seen

JRoy committed:
Removed unused sub class

JRoy committed:
Remove legacy nickname code for versions < 1.8

mdcfe commented on Essentials/src/messages.properties:
```diff
@@ -244,6 +244,8 @@ jailReleased=\u00a76Player \u00a7c{0}\u00a76 unjailed.
 jailSentenceExtended=\u00a76Jail time extended to \u00a7c{0}\u00a76.
 jailSet=\u00a76Jail\u00a7c {0} \u00a76has been set.
+jumpEasterDisable=\u00a76Flying wizard mode disabled
+jumpEasterEnable=\u00a76Enabling flying wizard mode
```
```suggestion
jumpEasterDisable=\u00a76Flying wizard mode disabled.
jumpEasterEnable=\u00a76Flying wizard mode enabled.
```

JRoy committed:
reply to remove

JRoy committed:
Commands repair to thunder

JRoy committed:
Print cause exceptions for NotEnoughArguments exceptions

JRoy committed:
Fix spacing in tick formatting

JRoy committed:
Cleanup time command

JRoy committed:
Finish commands

JRoy committed:
Clean up module commands

JRoy committed:
Pepega Clap

Ichbinjoe commented:
for the love of god could i please review smaller diffs

Ichbinjoe commented:
for the love of god could i please review smaller diffs

mdcfe commented on Essentials/src/com/earth2me/essentials/CommandSource.java:
```diff
@@ -22,8 +22,15 @@ public final Player getPlayer() {
         return null;
     }
 
+    public final net.ess3.api.IUser getUser(IEssentials ess) {
```
This feels icky. Why move this to `CommandSource`?

JRoy replied:
Because `ess.getUser(source.getSender())` vs `source.getUser(ess)`

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandpaytoggle.java:
```diff
@@ -20,11 +20,7 @@ public void run(Server server, User user, String commandLabel, String[] args) th
-        } else {
-            user.sendMessage(tl("payToggleOff"));
-        }
+        user.sendMessage(acceptingPay ? tl("payToggleOn") : tl("payToggleOff"));
```
Might be worth turning this into an `EssentialsLoopCommand` (see #3087).

JRoy committed:
Merge branch '2.x' into feature/command-rewrite

pop4959 commented on Essentials/src/com/earth2me/essentials/IEssentials.java:
```diff
@@ -103,7 +103,7 @@
     Collection<Player> getOnlinePlayers();
 
-    Iterable<User> getOnlineUsers();
+    List<User> getOnlineUsers();
```
Maybe `Collection<User>` would be better here?

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandclearinventory.java:
```diff
@@ -16,73 +16,55 @@
 
 
-public class Commandclearinventory extends EssentialsCommand {
+public class Commandclearinventory extends EssentialsLoopCommand {
```
Commands that have `*` as their first argument seem not to work, for example `/clear * **` or `/clear * dirt 1`.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -18,9 +18,6 @@
 
 
 public class Commandeco extends EssentialsLoopCommand {
```
`/eco give <player> -1` takes $1, but `/eco take pop4959 -1` does not give $1, which is a bit of an inconsistency in the command.

JRoy replied:
Just tested this, the `-` is ignored just like it way before this refactor

pop4959 replied:
I guess this is technically a "refactor", so you shouldn't have to change any preexisting behavior, but it's still strange that it works this way.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandeco.java:
```diff
@@ -32,82 +29,61 @@ public void run(final Server server, final CommandSource sender, final String co
     }
 
+    @Override
+    protected void updatePlayer(Server server, CommandSource sender, User user, String[] args) throws NotEnoughArgumentsException, PlayerExemptException, ChargeException, MaxMoneyException {
```
I think I prefer this over the consumer, and it's a bit weird to have only one implemented or the other

JRoy replied:
md wanted this to remain abstract https://github.com/EssentialsX/Essentials/pull/3337#discussion_r432844893

As for consumers vs this method, consumers allows us not to have to use instance/static variables laying around the command.

pop4959 replied:
Fair enough. I don't dislike the new method, but it could perhaps be its own distinct class (still abstract), since only one or the other method is ever being used.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandexp.java:
```diff
@@ -1,140 +1,136 @@
-            }
-            foundUser = true;
-            setExp(sender, player, amount, give);
+        boolean hasPermission(IUser user) {
```
This seems a little weird to me, why not just check the permissions normally but maybe just store the permission node on the enum?

JRoy replied:
This way it reduces code duplication

pop4959 replied:
Feels a bit forced but OK

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandjails.java:
```diff
@@ -14,7 +14,7 @@ public Commandjails() {
     @Override
     protected void run(final Server server, final CommandSource sender, final String commandLabel, final String[] args) throws Exception {
-        if (ess.getJails().getCount() < 1) {
+        if (ess.getJails().getCount() == 1) {
```
This is a bug

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandjump.java:
```diff
@@ -24,10 +24,10 @@ public void run(final Server server, final User user, final String commandLabel,
             if (user.isFlyClickJump()) {
                 user.setRightClickJump(false);
-                user.sendMessage("Flying wizard mode disabled");
+                user.sendMessage(tl("jumpEasterDisable"));
```
This easter egg doesn't appear to be functional. It's probably not worth adding to the translations.

JRoy replied:
/shrug

khobbits replied:
Why wouldn't this be working? :P
Steps:

- Enable Fly
- Fly in the sky
- Type /jump lock
- punch the air, while looking at the ground some distance away

You will be teleported to the location x/z, but keep the same y.

Odd feature... But that's why it was an easteregg.
Basically

pop4959 replied:
I stand corrected! I wasn't in fly mode, and I must have glossed over that when looking at the feature in code. Either way, it's an easteregg so it being translated or not too consequential. I'm impartial to it, since the feature is indeed working.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -32,12 +32,12 @@ public void run(final Server server, final CommandSource sender, final String co
             }
         } else {
-            if (user.isAuthorized("essentials.mute.exempt") && sender.isPlayer()) {
+            if (user.isAuthorized("essentials.mute.exempt")) {
```
nitpick: could be condensed to `else if {`

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandnuke.java:
```diff
@@ -25,10 +25,8 @@ protected void run(final Server server, final CommandSource sender, final String
-            for (String arg : args) {
-                targets.add(getPlayer(server, sender, args, pos).getBase());
-                pos++;
+            for (int i = 0; args.length > i; i++) {
```
```suggestion
            for (int i = 0; i < args.length; ++i) {
```

pop4959 replied:
In response to the commit message `pop wanted this idk why tho`: The way I've suggested it is more commonly used (I don't think I've seen it the other way often if at all in Essentials, or most other code). Late night me reviewing this thought this was a bug at first because it was the other way around.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandptime.java:
```diff
@@ -1,21 +1,20 @@
-public class Commandptime extends EssentialsCommand {
-    private static final Set<String> getAliases = new HashSet<>();
+public class Commandptime extends EssentialsLoopCommand {
+    protected static final Set<String> getAliases = new HashSet<>();
```
This should remain private. Also, perhaps `<T> Arrays::asList(T...)` would be cleaner here instead of the set + static block.

JRoy replied:
Is protected for its use in Commandpweather

pop4959 replied:
Only because you've forced it to be used in Commandpweather. Is there any particular reason why this is necessary? I think encapsulating each command is more important, and there is no other command as far as I know that borrows code from another directly like this.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandpweather.java:
```diff
@@ -1,27 +1,21 @@
 import java.util.*;
 
 import static com.earth2me.essentials.I18n.tl;
+import static com.earth2me.essentials.commands.Commandptime.getAliases;
```
Big nope

pop4959 replied:
I noticed this after seeing the `protected` scope change earlier (hence why I asked why you did it). Now I see, but this isn't a good idea in my opinion.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandrtoggle.java:
```diff
@@ -22,7 +22,7 @@ public void run(Server server, CommandSource sender, String commandLabel, String
 
     @Override
-    void togglePlayer(CommandSource sender, User user, Boolean enabled) throws NotEnoughArgumentsException {
+    protected void togglePlayer(CommandSource sender, User user, Boolean enabled) {
```
Why protected?

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandspeed.java:
```diff
@@ -133,18 +131,9 @@ private float getRealMoveSpeed(final float userSpeed, final boolean isFly, final
-    }
-
     @Override
     protected List<String> getTabCompleteOptions(Server server, CommandSource sender, String commandLabel, String[] args) {
```
This tab method is a little weird, because it doesn't actually suggest anything until the user has entered 2 args already, making it fairly useless currently.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandthunder.java:
```diff
@@ -18,22 +18,21 @@ public Commandthunder() {
-
-        } else {
+        if (args.length == 1) {
             world.setThundering(setThunder);
```
This method doesn't need to be called twice

JRoy replied:
Which method?

pop4959 replied:
Ignore, was blind and didn't see the return.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandtime.java:
```diff
@@ -166,25 +130,29 @@ private String normalizeWorldName(World world) {
+            if (args[0].equalsIgnoreCase("set")) {
+                return Lists.newArrayList("sunrise", "day", "morning", "noon", "afternoon", "sunset", "night", "midnight");
+            } else if (args[0].equalsIgnoreCase("add")) {
+                return Lists.newArrayList("500", "1000", "1500", "2000", "2500", "3000");
```
Numbers are a bit weird to autocomplete (they are ordered in lexicographical order), so maybe leave out 500 and increment by 1000 or something, or just leave out completions entirely.

pop4959 commented on Essentials/src/com/earth2me/essentials/commands/Commandtime.java:
```diff
@@ -166,25 +130,29 @@ private String normalizeWorldName(World world) {
     protected List<String> getTabCompleteOptions(Server server, CommandSource sender, String commandLabel, String[] args) {
-        final User user = ess.getUser(sender.getPlayer());
-
+        boolean isConsole = sender.isPlayer();
```
This logic doesn't seem right to me

JRoy replied:
What logic?

pop4959 replied:
if the sender is a player, they are not console.

JRoy replied:
oh lmao, fixed i fixed that up

pop4959 commented on Essentials/src/messages.properties:
```diff
@@ -772,6 +776,7 @@ timeCommandUsage=/<command> [day|night|dawn|17:30|4pm|4000ticks] [worldname|all]
 timeFormat=\u00a7c{0}\u00a76 or \u00a7c{1}\u00a76 or \u00a7c{2}\u00a76
 timeSetPermission=\u00a74You are not authorized to set the time.
 timeSetWorldPermission=\u00a74You are not authorized to set the time in world ''{0}''.
+timeWorldAdd=\u00a76The time was increased by\u00a7c {0} \u00a76in\: \u00a7c{1}\u00a76.
```
Maybe better: "Time was moved forward by..."

pop4959 commented on Essentials/src/plugin.yml:
```diff
@@ -621,9 +621,6 @@ permissions:
   essentials.silentquit:
     default: false
     description: Suppress leave/quit messages for users with this permission.
-  essentials.sleepingignored:
```
Why is this getting removed?

JRoy replied:
It was a duplicate entry

pop4959 replied:
Ah I see

pop4959 commented on Essentials/src/plugin.yml:
```diff
@@ -438,7 +438,7 @@ commands:
   time:
     description: Display/Change the world time. Defaults to current world.
-    usage: /<command> [day|night|dawn|17:30|4pm|4000ticks] [worldname|all]
+    usage: /<command> [set|add] <day|night|dawn|17:30|4pm|4000ticks> [worldname|all]
```
Optional arguments should use square brackets

pop4959 commented on Essentials/src/plugin.yml:
```diff
@@ -133,8 +133,8 @@ commands:
   exp:
-    description: Give, set or look at a players exp.
-    usage: /<command> [show|set|give] [playername [amount]]
+    description: Give, set, reset, or look at a players exp.
```
These descriptions/usage will need to be updated in messages.properties as well

mdcfe replied:
```suggestion
    description: Give, set, reset, or look at a players experience.
```

pop4959 commented on EssentialsXMPP/src/com/earth2me/essentials/xmpp/Commandxmpp.java:
```diff
@@ -21,13 +21,14 @@ protected void run(final Server server, final CommandSource sender, final String
+        final String senderName = sender.isPlayer() ? ess.getUser(sender.getPlayer()).getDisplayName() : Console.NAME;
+        sender.sendMessage("[" + senderName + ">" + address + "] " + message);
+        if (!EssentialsXMPP.getInstance().sendMessage(address, "[" + senderName + "] " + message)) {
+            sender.sendMessage("§cError sending message.");
```
Would probably be better to use ChatColor here

pop4959 reviewed (changes requested):
Not a thorough review by any means, there is quite a lot here to unpack. Seems like most things have been tested (or at least I hope so), so that's good. One thing I'm quite iffy about are some of the changes to EssentialsLoopCommand (in particular, leaving empty method stubs). I can see how the refactored version can be useful, but maybe it should be its own separate class.

JRoy committed:
Merge branch '2.x' into feature/command-rewrite

JRoy committed:
Make IEssentials#getOnlineUsers return Collection

JRoy committed:
Fix Commandclearinventory offsets

JRoy committed:
Fix bug with only one jail defined

JRoy committed:
Remove TODO

JRoy committed:
Fix nested else if to elseif

JRoy committed:
pop wanted this idk why tho

JRoy committed:
Convert getAliases from a Set to a List

JRoy committed:
Remove protected status for Commandrtoggle#togglePlayer

JRoy committed:
Add tab completion for speed command

JRoy committed:
Optimize Commandtime command completions

JRoy committed:
Remove unnecessary console check

JRoy committed:
Change timeWorldAdd message

JRoy committed:
Update command usages/descriptions in messages.properties

JRoy committed:
Stupid legacy module doesn't use stupid chat color enum

JRoy committed:
Remove protected status from Commandptime#getAliases

pop4959 reviewed (dismissed):
More or less seems OK, I'd recommend someone else also review this to find anything I might've missed.

mibby commented:
Probably not the appropriate place for this, but thought I'd let you know Jroy that there are merge conflicts with this PR (3337) and the reduced sync load PR (3102). Was going around testing a bunch of fixes and improvements from both you and pop4959.

JRoy commented:
@mibby i would assume there to be, I'll update this once async pr merged

pop4959 commented:
@mibby Thanks for helping test; appreciate it!

JRoy committed:
Merge branch '2.x' into feature/command-rewrite

# Conflicts:
#	Essentials/src/com/earth2me/essentials/User.java
#	Essentials/src/com/earth2me/essentials/commands/Commandback.java
#	Essentials/src/com/earth2me/essentials/commands/Commandtogglejail.java
#	EssentialsSpawn/src/com/earth2me/essentials/spawn/Commandspawn.java

JRoy commented:
Merge conflicts resolved

JRoy committed:
Merge branch '2.x' into feature/command-rewrite

# Conflicts:
#	Essentials/src/com/earth2me/essentials/Essentials.java

JRoy committed:
Merge branch '2.x' into feature/command-rewrite

JRoy commented:
Merge conflicts resolved

JRoy committed:
Merge branch '2.x' into feature/command-rewrite

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandspeed.java:
```diff
@@ -6,13 +6,17 @@
 
 
 public class Commandspeed extends EssentialsCommand {
+    private final List<String> types = Arrays.asList("walk", "fly", "1", "1.5", "1.75", "2");
+    private final List<String> speeds = Arrays.asList("1", "1.5", "1.75", "2");
```
```suggestion
    private static final List<String> types = Arrays.asList("walk", "fly", "1", "1.5", "1.75", "2");
    private static final List<String> speeds = Arrays.asList("1", "1.5", "1.75", "2");
```

pop4959 replied:
I would also suggest using a better range of speeds, such as 1,2,3,4,5 instead of these random decimals

pop4959 replied:
or just maybe don't put the numbers lol its kind of weird to suggest specific numbers in general imo

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandtime.java:
```diff
@@ -14,120 +14,94 @@
+        }
     }
 
-    /**
-     * Used to parse an argument of the type "world(s) selector"
-     */
```
Comment should probably be improved not just removed

mdcfe commented on Essentials/src/messages.properties:
```diff
@@ -191,8 +191,8 @@ essentialsHelp1=The file is broken and Essentials can''t open it. Essentials is
 exp=\u00a7c{0} \u00a76has\u00a7c {1} \u00a76exp (level\u00a7c {2}\u00a76) and needs\u00a7c {3} \u00a76more exp to level up.
-expCommandDescription=Give, set or look at a players exp.
-expCommandUsage=/<command> [show|set|give] [playername [amount]]
+expCommandDescription=Give, set, reset, or look at a players exp.
```
```suggestion
expCommandDescription=Give, set, reset, or look at a players experience.
```

mdcfe commented on EssentialsSpawn/src/com/earth2me/essentials/spawn/Commandsetspawn.java:
```diff
@@ -18,4 +22,9 @@ public void run(final Server server, final User user, final String commandLabel,
+
+    @Override
+    protected List<String> getTabCompleteOptions(Server server, CommandSource sender, String commandLabel, String[] args) {
+        return Collections.emptyList();
```
Should return a list of groups in the permission manager.

JRoy committed:
Update Essentials/src/messages.properties

Co-authored-by: MD <1917406+md678685@users.noreply.github.com>

JRoy committed:
Update Essentials/src/plugin.yml

Co-authored-by: MD <1917406+md678685@users.noreply.github.com>

JRoy committed:
Update Essentials/src/com/earth2me/essentials/commands/Commandspeed.java

Co-authored-by: MD <1917406+md678685@users.noreply.github.com>

JRoy committed:
Update Essentials/src/messages.properties

Co-authored-by: MD <1917406+md678685@users.noreply.github.com>

JRoy committed:
Comment

JRoy committed:
Move paytoggle to ToggleCommand

JRoy committed:
Merge branch '2.x' into feature/command-rewrite

# Conflicts:
#	Essentials/src/com/earth2me/essentials/Essentials.java
#	Essentials/src/com/earth2me/essentials/commands/Commandseen.java

JRoy committed:
Fix #3579

JRoy commented:
EZ Clap

NicoNekoDev commented:
Finally. 😎
